### PR TITLE
[TEP-0144] Validate PipelineRun for Param Enum

### PR DESCRIPTION
**Pull request by QuanZhang-William**
# Changes

Part of [#7270][#7270]. In [TEP-0144][tep-0144] we proposed a new `enum` field to support built-in param input validation.

This commit adds validation logic for PipelineRun against Param Enum

/kind feature

[#7270]: https://github.com/tektoncd/pipeline/issues/7270
[tep-0144]: https://github.com/tektoncd/community/blob/main/teps/0144-param-enum.md

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Implement Param Enum validation for PipelineRuns. Param Enum is supported per TEP-0144
```

### CODE REVIEW

File: docs/pipelineruns.md
Comment (Yongxuanzhang) [inline docs/pipelineruns.md:279]: Is this PR the last PR?

File: examples/v1/pipelineruns/alpha/param-enum.yaml
Comment (Yongxuanzhang) [inline examples/v1/pipelineruns/alpha/param-enum.yaml:22]: optional, maybe validate that the $(params.message) is v2?

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/pipelinerun.go:393]: Is it possible that the Conditions[0] is not the condition we need? We can loop the conditions to fetch the apisucceed condition
And curious if we have a permanent error from taskrun, is it surfaced somewhere?

File: docs/pipelineruns.md
Comment (QuanZhang-William) [reply docs/pipelineruns.md:279]: Yeah, it is. I was planning to do the doc and link cleanup in a separate PR but seems unncessary. Will include the cleanup in this PR

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go:393]: Oh, good point. If a Matrix is used and multiple TaskRun is created with different combintations. We do need to validate all the results.

And yes, we do have a permanent error from TaskRun, which is also Propogated back to the owning `PipelineRun`:
https://github.com/tektoncd/pipeline/blob/7e3897a6353f81836e2f9186259bb78c21434b20/pkg/reconciler/taskrun/taskrun.go#L420C1-L426C3

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/pipelinerun.go:394]: is it possible that Conditions[0] is not `apis.ConditionSucceeded` ?

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/pipelinerun.go:397]: where does the err in controller.NewPermanentError(err) come from? Looks like it is created in previous code, but not from this code

Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/pipelinerun.go]: Sorry.. I think I pushed too fast. Probably need to take a rest for the day 😵

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/pipelinerun.go]: 😄

**Review by Yongxuanzhang — APPROVED** [review]

**Comment by QuanZhang-William** [comment]
@tektoncd/core-collaborators @tektoncd/core-maintainers. Please take a look if you are interested 😃

File: docs/pipelineruns.md
Comment (jerop) [inline docs/pipelineruns.md:276]:
```suggestion
> :seedling: **`enum` is an [alpha](additional-configs.md#alpha-features) feature.** The `enable-param-enum` feature flag must be set to `"true"` to enable this feature.
```

File: docs/pipelineruns.md
Comment (jerop) [inline docs/pipelineruns.md:280]:
```suggestion
Tekton will also the validate the `param` values passed to any referenced `Tasks` (via `taskRef`) if `Enum` is specified for the `Task`. The `PipelineRun` will fail with reason `InvalidParamValue` if `Enum` validation is failed for any of the `PipelineTask`. 
```

File: docs/pipelineruns.md
Comment (jerop) [inline docs/pipelineruns.md:282]:
```suggestion
You can also specify `Enum` an embedded `Pipeline` in a `PipelineRun`. The same `param` validation will be executed in this scenario.
```

File: docs/pipelines.md
Comment (jerop) [inline docs/pipelines.md:275]:
```suggestion
> :seedling: **`enum` is an [alpha](additional-configs.md#alpha-features) feature.** The `enable-param-enum` feature flag must be set to `"true"` to enable this feature.
```

File: docs/pipelines.md
Comment (jerop) [inline docs/pipelines.md:303]: apologies for coming to this late, i wasn't involved in the design discussions

i find this behavior to be confusing -- for example, as a user of a pipeline, I'd be very confused why a pipelinerun failed when `v3` is listed as a valid value for the pipeline

what if we require that if both pipeline and task specify an enum, then the enum in the pipeline must be a subset of the enum in the task? that way, the issue is with the pipeline specification and not an issue that users run into occasionally even when they pass in certain values in the pipeline param enum

**Review by jerop — CHANGES_REQUESTED** [review]
I have a concern about the intersection approach when `enum` is specified in both tasks and pipelines, I think we should require that the `enum` in pipeline is a subset of the `enum` in task

Also, what happens currently when the param in pipeline is passed to multiple tasks with enums? do you take the intersection of the enums in all the tasks the param is passed to and the pipeline? in the example below, what is the valid value for "message" param in a pipelinerun?

``` yaml
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: param-enum-demo-1
spec:
  params:
  - name: message
    type: string
    enum: ["v1", "v2", "v3"]
  steps:
  - name: build
    image: bash:latest
    script: |
      echo "$(params.message)"
---
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: param-enum-demo-2
spec:
  params:
  - name: message
    type: string
    enum: ["v3", "v4"]
  steps:
  - name: build
    image: bash:latest
    script: |
      echo "$(params.message)"
---
apiVersion: tekton.dev/v1
kind: Pipeline
metadata:
  name: pipeline-param-enum
spec:
  params:
  - name: message
    enum: ["v1", "v3", "v4", "v5"]
  tasks:
  - name: task-1
    params:
      - name: message
        value: $(params.message)
    taskRef:
      name: param-enum-demo-1
  - name: task-2 
    params:
      - name: message
        value: $(params.message)
    taskRef:
      name: param-enum-demo-2
```

**Comment by jerop** [comment]
/hold

wait on https://github.com/tektoncd/community/pull/1106 to be merged

File: docs/pipelines.md
Comment (QuanZhang-William) [reply docs/pipelines.md:303]: Thanks @jerop for the suggestion. I've opened a TEP update to reflect what you suggested: https://github.com/tektoncd/community/pull/1106

Comment (QuanZhang-William) [reply docs/pipelines.md]: /hold cancel
https://github.com/tektoncd/community/pull/1106 is merged.

Based on the discussion in the TEP, I have added the subset validation, PTAL!

**Comment by QuanZhang-William** [comment]
> I have a concern about the intersection approach when `enum` is specified in both tasks and pipelines, I think we should require that the `enum` in pipeline is a subset of the `enum` in task
> 
> Also, what happens currently when the param in pipeline is passed to multiple tasks with enums? do you take the intersection of the enums in all the tasks the param is passed to and the pipeline? in the example below, what is the valid value for "message" param in a pipelinerun?
> 
> ```yaml
> apiVersion: tekton.dev/v1
> kind: Task
> metadata:
>   name: param-enum-demo-1
> spec:
>   params:
>   - name: message
>     type: string
>     enum: ["v1", "v2", "v3"]
>   steps:
>   - name: build
>     image: bash:latest
>     script: |
>       echo "$(params.message)"
> ---
> apiVersion: tekton.dev/v1
> kind: Task
> metadata:
>   name: param-enum-demo-2
> spec:
>   params:
>   - name: message
>     type: string
>     enum: ["v3", "v4"]
>   steps:
>   - name: build
>     image: bash:latest
>     script: |
>       echo "$(params.message)"
> ---
> apiVersion: tekton.dev/v1
> kind: Pipeline
> metadata:
>   name: pipeline-param-enum
> spec:
>   params:
>   - name: message
>     enum: ["v1", "v3", "v4", "v5"]
>   tasks:
>   - name: task-1
>     params:
>       - name: message
>         value: $(params.message)
>     taskRef:
>       name: param-enum-demo-1
>   - name: task-2 
>     params:
>       - name: message
>         value: $(params.message)
>     taskRef:
>       name: param-enum-demo-2
> ```

Hi Jerop. You are right about the behaviour in the previous intersection design, it essentially takes the intersection of all related enum sets. I have changed to subset design, which also essentially validates the pipeline-level is a subset of all referenced Tasks

File: docs/pipelines.md
Comment (jerop) [inline docs/pipelines.md:288]: do we have validation that the default value is in the enum? mind pointing me to it?

it may be obvious, but please let's add docs making it clear that it's not just that values passed to the pipelinerun needs to be in the enum but also the default value

Comment (QuanZhang-William) [reply docs/pipelines.md:288]: Hi Jerop, yes we do: 
https://github.com/tektoncd/pipeline/blob/140b633d8e9ca07f81f236a5a234c8665f7058ca/pkg/apis/pipeline/v1/param_types.go#L165C1-L169C4

I will update the doc to make it explicit

Comment (QuanZhang-William) [reply docs/pipelines.md:288]: Doc updated for both pipelines and tasks, PTAL.

**Review by jerop — COMMENTED** [review]
Thank you @QuanZhang-William!
 
/lgtm

**Pull request merged**
